src/rest-server/src/middlewares/v2/protocol.js

```diff
@@ -0,0 +1,154 @@
+const yaml = require('js-yaml');
+const mustache = require('mustache');
+const createError = require('../../util/error');
+const protocolSchema = require('../../config/v2/protocol');
+
+
+const prerequisiteTypes = [
+  'script',
+  'output',
+  'data',
+  'dockerimage',
+];
+
+const prerequisiteFields = [
+  'script',
+  'output',
+  'data',
+  'dockerImage',
+];
+
+const protocolValidate = (protocolYAML) => {
+  const protocolObj = yaml.safeLoad(protocolYAML);
+  if (!protocolSchema.validate(protocolObj)) {
+    throw createError('Bad Request', 'InvalidProtocolError', protocolSchema.validate.errors);
+  }
+  // convert prerequisites list to dict
+  const prerequisites = {};
+  for (let type of prerequisiteTypes) {
+    prerequisites[type] = {};
+  }
+  if ('prerequisites' in protocolObj) {
+    for (let item of protocolObj.prerequisites) {
+      prerequisites[item.type][item.name] = item;
+    }
+  }
+  protocolObj.prerequisites = prerequisites;
+  // convert deployments list to dict
+  const deployments = {};
+  if ('deployments' in protocolObj) {
+    for (let item of protocolObj.deployments) {
+      deployments[item.name] = item;
+    }
+  }
+  protocolObj.deployments = deployments;
+  // check prerequisites in taskRoles
+  for (let taskRole of Object.keys(protocolObj.taskRoles)) {
+    for (let field of prerequisiteFields) {
+      if (field in protocolObj.taskRoles[taskRole] &&
+        !(protocolObj.taskRoles[taskRole][field] in prerequisites[field.toLowerCase()])) {
+        throw createError(
+          'Bad Request',
+          'InvalidProtocolError',
+          `Prerequisite ${protocolObj.taskRoles[taskRole][field]} does not exist.`
+        );
+      }
+    }
+  }
+  // check deployment in defaults
+  if ('defaults' in protocolObj) {
+    if ('deployment' in protocolObj.defaults &&
+      !(protocolObj.defaults.deployment in deployments)) {
+        throw createError(
+          'Bad Request',
+          'InvalidProtocolError',
+          `Default deployment ${protocolObj.defaults.deployment} does not exist.`
+        );
+    }
+  }
+  return protocolObj;
+};
+
+const protocolRender = (protocolObj) => {
+  let deployment = null;
+  if ('defaults' in protocolObj && 'deployment' in protocolObj.defaults) {
+    deployment = protocolObj.deployments[protocolObj.defaults.deployment];
+  }
+  for (let taskRole of Object.keys(protocolObj.taskRoles)) {
+    let commands = protocolObj.taskRoles[taskRole].commands;
+    if (deployment != null && taskRole in deployment.taskRoles) {
+      if ('preCommands' in deployment.taskRoles[taskRole]) {
+        commands = deployment.taskRoles[taskRole].preCommands.concat(commands);
+      }
+      if ('postCommands' in deployment.taskRoles[taskRole]) {
+        commands = commands.concat(deployment.taskRoles[taskRole].postCommands);
+      }
+    }
+    let entrypoint = '';
+    commands = commands.map((command) => command.trim()).join(' && ');
+    const tokens = mustache.parse(commands, ['<%', '%>']);
```

Comment:
There is a bug in the version of mustache we are using is that the cache key of the parsed template is only the template source, w/o the delimiters, so if we are using different delimiters in the same cache, there may be parsing mistake occurs when parsing the same text with different delimiters.

Reply:
Updated.